### PR TITLE
Zendoo integration all with snark data

### PR DESCRIPTION
**Pull request by oaleksandrov-ip**

### CODE REVIEW

File: qa/SidechainTestFramework/sc_boostrap_info.py [hunk 1]
```diff
@@ -122,12 +142,13 @@ def __init__(self, secret, publicKey):
     "network":
     "withdrawal_epoch_length":
     "genesis_vrf_account": an instance of VrfAccount
+    "withdrawalCertificateData": an instance of WithdrawalCertificateData
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply qa/SidechainTestFramework/sc_boostrap_info.py:145]: done

File: qa/SidechainTestFramework/sc_boostrap_info.py [hunk 2]
```diff
@@ -122,12 +142,13 @@ def __init__(self, secret, publicKey):
...
 }
 """
 class SCBootstrapInfo(object):
 
     def __init__(self, sidechain_id, genesis_account, genesis_account_balance, mainchain_block_height,
-                 sidechain_genesis_block_hex, pow_data, network, withdrawal_epoch_length, genesis_vrf_account):
+                 sidechain_genesis_block_hex, pow_data, network, withdrawal_epoch_length, genesis_vrf_account, withdrawal_certificate_data=WithdrawalCertificateData):
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply qa/SidechainTestFramework/sc_boostrap_info.py:151]: done

File: qa/SidechainTestFramework/scutil.py [hunk 3]
```diff
@@ -118,12 +118,13 @@ def generate_genesis_data(genesis_info, genesis_secret, vrf_secret):
     jsonParameters = {"secret": genesis_secret, "vrfSecret": vrf_secret, "info": genesis_info}
 
     json_param = json.dumps(jsonParameters)
-
+    print json_param
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply qa/SidechainTestFramework/scutil.py:121]: done

File: qa/SidechainTestFramework/scutil.py [hunk 4]
```diff
@@ -118,12 +118,13 @@ def generate_genesis_data(genesis_info, genesis_secret, vrf_secret):
...
     javaPs = subprocess.Popen(["java", "-cp",
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply qa/SidechainTestFramework/scutil.py:122]: yes :) moreover could be extracted to separate function

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 5]
```diff
@@ -46,6 +46,7 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
   override type NVCT = SidechainState
 
   val sysDataConstant: Array[Byte] = CryptoLibProvider.sigProofThresholdCircuitFunctions.generateSysDataConstant(params.signersPublicKeys.map(_.bytes()).asJava, params.signersThreshold)
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:48]: done

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 6]
```diff
@@ -143,8 +146,11 @@ class CertificateSubmitter
             params)
 
           try {
-            log.info(s"Backward transfer certificate was successfully created. Cert hash = " +
-              BytesUtils.toHexString(mainchainApi.sendCertificate(certificate).certificateId))
+            log.info(s"Backward transfer certificate request was successfully created fo epoch number ${certificateRequest.epochNumber}, with proof ${BytesUtils.toHexString(proofWithQuality.getKey)} with quality ${proofWithQuality.getValue} try to send it to mainchain")
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:149]: done

File: sdk/src/test/scala/com/horizen/block/CertificateTest.scala [hunk 7]
```diff
@@ -7,16 +7,20 @@ import org.junit.Assert._
...
 
   @Test
   def emptyCertificate(): Unit = {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/test/scala/com/horizen/block/CertificateTest.scala:17]: There can be only one :)

File: tools/sctool/src/main/java/com/horizen/CommandProcessor.java [hunk 8]
```diff
@@ -197,19 +197,33 @@ private void processGenerateSchnorrKeys(JsonNode json) {
             return;
         }
 
+        //@TODO remove hardcoded value after supporting various verification/proving files, actual file is generated for 7 keys
+        if (keyCount != 7) {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply tools/sctool/src/main/java/com/horizen/CommandProcessor.java:201]: proof files are expected 7 keys

File: sdk/src/main/java/com/horizen/secret/SchnorrSecret.java [hunk 9]
```diff
@@ -12,6 +12,7 @@
 
 import static com.horizen.secret.SecretsIdsEnum.SchnorrSecretKeyId;
 
+//@TODO add JSON serialization
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/secret/SchnorrSecret.java:15]: Do we need it?

File: sdk/src/main/java/com/horizen/cryptolibprovider/ThresholdSignatureCircuitImplZendoo.java [hunk 10]
```diff
@@ -91,4 +87,15 @@ public int sysDataConstantLength() {
...
+
+    @Override
+    public int sysDataConstantLength() {
+        return PoseidonHash.HASH_LENGTH;
+    }
+
+    @Override
+    public int proofSizeLength() {
+        return 771; //@TODO take it from JNI side
+    }
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/cryptolibprovider/ThresholdSignatureCircuitImplZendoo.java:99]: Add also the method for VkCertKey length form sidechain creation output

File: sdk/src/main/java/com/horizen/transaction/mainchain/SidechainCreation.java [hunk 11]
```diff
@@ -64,7 +58,7 @@ public ForgerBox getBox() {
     }
 
     public byte[] getBackwardTransferPoseidonRootHash() {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/java/com/horizen/transaction/mainchain/SidechainCreation.java:60]: @oaleksandrov-ip what do you think?

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 12]
```diff
@@ -91,9 +89,12 @@ class CertificateSubmitter
     val actualSysDataConstant =
       CryptoLibProvider.sigProofThresholdCircuitFunctions.generateSysDataConstant(signersPublicKeys.map(_.bytes()).asJava, params.signersThreshold)
 
-    val expectedSysDataConstant = genSysConstant//getSidechainCreationTransaction(sidechainNodeView.history).getBackwardTransferPoseidonRootHash
+    val expectedSysDataConstant = params.calculatedSysDataConstant
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:92]: I think, that here we can use `params.calculatedSysDataConstant` directly, same as in State.

File: sdk/src/main/scala/com/horizen/SidechainApp.scala [hunk 13]
```diff
@@ -115,8 +119,8 @@ class SidechainApp @Inject()
       signersPublicKeys = signersPublicKeys,
       signersThreshold = sidechainSettings.backwardTransferSettings.signersThreshold,
       provingKeyFilePath = sidechainSettings.backwardTransferSettings.provingKeyFilePath,
-      verificationKeyFilePath = sidechainSettings.backwardTransferSettings.verificationKeyFilePath
-
+      verificationKeyFilePath = sidechainSettings.backwardTransferSettings.verificationKeyFilePath,
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainApp.scala:122]: done

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 14]
```diff
@@ -45,7 +45,7 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
 
   override type NVCT = SidechainState
 
-  val sysDataConstant: Array[Byte] = CryptoLibProvider.sigProofThresholdCircuitFunctions.generateSysDataConstant(params.signersPublicKeys.map(_.bytes()).asJava, params.signersThreshold)
+  val sysDataConstant: Array[Byte] = params.calculatedSysDataConstant
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:48]: Done, but it is better not to expose the source of data. Anyway we everywhere use direct params value, so changed

File: sdk/src/main/scala/com/horizen/block/MainchainTxSidechainCreationCrosschainOutput.scala [hunk 15]
```diff
@@ -49,15 +49,15 @@ object MainchainTxSidechainCreationCrosschainOutput {
...
-    val constant: Array[Byte] = new Array[Byte](constantLength) // TODO: uncomment when possible
-    //val constant: Array[Byte] = sidechainCreationOutputBytes.slice(currentOffset, currentOffset + constantLength)
-    //currentOffset += constantLength
+    val constantLength: VarInt = BytesUtils.getReversedVarInt(sidechainCreationOutputBytes, currentOffset)
+    currentOffset += constantLength.size()
+
+    val constant: Array[Byte] = sidechainCreationOutputBytes.slice(currentOffset, currentOffset + constantLength.value().intValue())
+    currentOffset += constantLength.value().intValue()
 
     val certVkSize: Int = 1544 // TODO: take it from zendoo interface
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/block/MainchainTxSidechainCreationCrosschainOutput.scala:58]: done

File: sdk/src/main/java/com/horizen/secret/SchnorrSecret.java (hunk 9 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/secret/SchnorrSecret.java:15]: yes, for example for bootstrapping tool

File: sdk/src/main/java/com/horizen/cryptolibprovider/ThresholdSignatureCircuitImplZendoo.java (hunk 10 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/cryptolibprovider/ThresholdSignatureCircuitImplZendoo.java:99]: done

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala (hunk 12 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:92]: changed

File: changelog.md [hunk 16]
```diff
@@ -0,0 +1,22 @@
+**Beta features**
+1. Multiple sidechain nodes network.
+2. Mainchain synchronization: Cross-chain Transfer Protocol support for sidechain declaration: forward and backward transfers.
+3. Proof-of-Stake consensus protocol for sidechain based on Ouroboros Praos consensus protocol.
```
Comment (i-Alex) [inline changelog.md:4]: Let's specify the name of the protocol like `Latus - Proof-of-Stake ....` and put the link to whitepaper 2.0

File: changelog.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
+**Beta features**
+1. Multiple sidechain nodes network.
+2. Mainchain synchronization: Cross-chain Transfer Protocol support for sidechain declaration: forward and backward transfers.
```
Comment (i-Alex) [inline changelog.md:3]: 2. Mainchain synchronization: Cross-chain Transfer Protocol support for sidechain declaration, forward transfers, backward transfer requests and withdrawal certificates.
3. Withdrawal certificate with threshold signature zero-knowledge proof (link to zendoo-sc-cryptolib)

File: changelog.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+3. Proof-of-Stake consensus protocol for sidechain based on Ouroboros Praos consensus protocol.
+4. Built-in coins transferring operations inside sidechain.
```
Comment (i-Alex) [inline changelog.md:5]: Let's specify the new kind of transaction - SidechainCoreTranscation and also spesify the Boxes structure was improved/changed.

File: changelog.md [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+5. HTTP API for basic node operations.
+6. Possibility to declare custom Transactions/Boxes/Secrets/etc.
+7. Possibility to extend/manage basic API.
+8. Web interface and command line tool for interaction with the Node.
+9. Sidechain Bootstapping Tool to configure sidechain network according to the mainchain network.
+
```
Comment (i-Alex) [inline changelog.md:11]: 10. Mainchain caused forks processing by Forger, automatic forging, etc.

File: changelog.md [hunk 20]
```diff
@@ -0,0 +1,22 @@
+**Beta features**
```
Comment (i-Alex) [inline changelog.md:1]: All the comments below, can be done in main readme.md for Beta description as well.
To discuss:
1. Do we need to call it `beta` or `beta preview`?
2. Do we need to keep in changelog the same list as in readme.md, or only the real changes? Moreover in changes we can have extended list with different changes, but in readme.md just most important features supported for now.

File: examples/simpleapp/readme.md [hunk 21]
```diff
@@ -1,7 +1,7 @@
 **Simple App**
 ---------
 
-Simple app is a Java application that allows to run a pure Sidechain Node without any custom logic.
+Simple app is a Java application that allows to run a pure Sidechain Node without any custom logic like custom boxes and transactions. Forward and Backward transactions are available from box.
```
Comment (i-Alex) [inline examples/simpleapp/readme.md:4]: `Forward and Backward transactions are available from box.` -> `Forward and Backward transfers are available out of the box.`

File: readme.md [hunk 22]
```diff
@@ -20,7 +22,7 @@ Sidechains-SDK is available and tested on Linux and Windows (64bit).
 
 **Requirements**
 
-Install Java 8 or newer, Scala 2.12.x, SBT 0.13.x
+Install Java 11 or newer, Scala 2.12.x, SBT 0.13.x
```
Comment (i-Alex) [inline readme.md:25]: As I remember, min version is Java 8

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala [hunk 23]
```diff
@@ -86,10 +86,9 @@ class CertificateSubmitter
   private def checkSubmitterMessage(sidechainNodeView: View): Try[Unit] = Try {
     val signersPublicKeys = params.signersPublicKeys
 
-    val actualSysDataConstant =
-      CryptoLibProvider.sigProofThresholdCircuitFunctions.generateSysDataConstant(signersPublicKeys.map(_.bytes()).asJava, params.signersThreshold)
+    val actualSysDataConstant = params.calculatedSysDataConstant
+    val expectedSysDataConstant = getSidechainCreationTransaction(sidechainNodeView.history).getBackwardTransferPoseidonRootHash
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:90]: `getBackwardTransferPoseidonRootHash` name is outdated. See the comment there.

File: examples/simpleapp/readme.md [hunk 24]
```diff
@@ -35,41 +35,26 @@ Why it is "isolated"? Because inside SC genesis block declared a MC Block Refere
...
-2. Run MC node in a `regtest` mode with a `-websocket` flag. 
-3. Generate at least 220 blocks to enable SC support.
-4. Declare the new SC using `sc_create` RPC method, where you will specify new SC id and first forward transfers.
-5. Generate 1 more block, that includes the transaction previously generated.
-6. Get SC genesis info from MC for your SC ID using `getscgenesisinfo` RPC method.
-7. Run ScBootstrapping tool and execute a command `genesisinfo`, where the result from previous RPC method should be placed as well as a secret, that will be used to sign the SC genesis block.
-8. Use the result data and update configuration file with it. You can use configuration files from `./examples/simpleapp/src/main/resourse/` as an example of resulting structure of the file.
-9. Run SimpleApp node with the configuration file you created. Your SC node will be connected to the MC node.
-10. If new MC blocks will appear, than the next SC blocks will automatically include the references to MC blocks inside. Maximum number of MC Block Refs per SC Block is 3.
+**Running SimpleApp connected to the real MC node**
```
Comment (i-Alex) [inline examples/simpleapp/readme.md:39]: I would prefer to leave a short version here too and make it actual.. Moreover here we specify the proper MC git branch and some additional data that is not present in `mc_sc_workflow_example.md`

File: readme.md [hunk 25]
```diff
@@ -40,4 +42,4 @@ Project has a module maven structure and consists of 3 modules:
 
 **Examples**
 
-To have a sidechain node without any custom business logic run [Simple App](examples/simpleapp/readme.md) example.
+To have a sidechain node without any custom business logic run [Simple App](examples/simpleapp/mc_sc_workflow_example.md) example.
```
Comment (i-Alex) [inline readme.md:45]: Why not to refer to `examples/simpleapp/readme.md` where the first Simple-app description is present?

File: changelog.md (hunk 17 above)
Comment (oaleksandrov-ip) [reply changelog.md:3]: done

File: changelog.md (hunk 16 above)
Comment (oaleksandrov-ip) [reply changelog.md:4]: done

File: changelog.md (hunk 19 above)
Comment (oaleksandrov-ip) [reply changelog.md:11]: done

File: examples/simpleapp/readme.md (hunk 21 above)
Comment (oaleksandrov-ip) [reply examples/simpleapp/readme.md:4]: done

File: readme.md (hunk 22 above)
Comment (oaleksandrov-ip) [reply readme.md:25]: changed

File: readme.md (hunk 25 above)
Comment (oaleksandrov-ip) [reply readme.md:45]: changed to have both references

File: sdk/src/main/java/com/horizen/transaction/mainchain/SidechainCreation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/transaction/mainchain/SidechainCreation.java:60]: renamed to `getGenSysConstant` we such name everywhere

File: sdk/src/main/scala/com/horizen/block/MainchainBlockReference.scala [hunk 26]
```diff
@@ -245,6 +247,7 @@ object MainchainBlockReference {
             offset += certificatesCount.size()
 
             while (certificates.size < certificatesCount.value()) {
+              log.info(s"Parse Mainchain certificate: ${BytesUtils.toHexString(util.Arrays.copyOfRange(mainchainBlockBytes, offset, mainchainBlockBytes.length))}")
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/block/MainchainBlockReference.scala:250]: changed to debug level

File: sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala (hunk 23 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala:90]: done

File: qa/readme.md [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+        ```
+
+**Template configuration files**
+
+Template configuration files are located in directory resources. 
+
+File template.conf is template for tests of SC node(s) is connected to MC node(s).
+
+File template_predefined_genesis.conf is template for test of SC node(s) in standalone mode(without connection to MC node).
+
```
Comment (i-Alex) [inline qa/readme.md:33]: We need to add also information about how to let STF to know the path to zend/zen-cli, boots tool, simpleapp

**Pull request merged**